libnmstate/nm/context.py

```diff
@@ -62,12 +100,92 @@ def _del_client(self):
                 timeout_source = GLib.timeout_source_new(50)
                 try:
                     timeout_source.set_callback(lambda x: is_timeout.append(1))
-                    timeout_source.attach(self._context)
+                    timeout_source.attach(self.context)
                     while not is_done and not is_timeout:
-                        self._context.iteration(True)
+                        self.context.iteration(True)
                 finally:
                     timeout_source.destroy()
-
             if not is_done:
                 logging.error("BUG: NM.Client is not cleaned")
             self._context = None
+
+    def _del_timeout(self):
+        if self._timeout_source:
+            self._timeout_source.destroy()
+            self._timeout_source = None
+
+    def register_async(self, action, fast=False):
+        """
+        Register action(string) to wait list.
+        """
+        queue = self._fast_queue if fast else self._slow_queue
+        max_queue = FAST_ASYNC_QUEUE_SIZE if fast else SLOW_ASYNC_QUEUE_SIZE
+        if len(queue) >= max_queue:
+            logging.debug(
+                f"Async queue({max_queue}) full, waiting all existing actions "
+                "to be finished before registering more async action"
+            )
+            self.wait_all_finish()
```

Comment:
IMO, it would be better to wait for a slot in queue than wait for all the actions to finish. Is not a must, but I think in some cases it could improve performance a little bit.

Reply:
This require me to create a `wait_one_finish()` with its own main context iteration code and that complex the flow.

Yes. That's a nice stuff to have. But the performance will not be a big difference.
We finish 200 interface activation in 18 seconds and most of time we are waiting NetworkManager. 

I have add a TODO comments there and we can work on it in another PR.